fmf/context.py

```diff
@@ -44,13 +47,13 @@ def __init__(self, origin):
         else:
             self._to_compare = self._split_to_version(origin)
 
-    def __eq__(self, other):
+    def __eq__(self, other: ContextValue):
```

Comment:
Linters will complain here: `ContextValue` is being defined right now, it is not yet fully known, therefore it cannot be used as an annotation directly, a forward reference is needed, `other: 'ContextValue'`.

Reply:
Also, note that the code does handle the possibility of `other` being something completely different than `ContextValue`, I'd say the right type here would be `Any` after all. If the `other` is recognized as `ContextValue` by `isinstance()` below, fine, it's compared, anything else leads to `False`.

Reply:
That would be an issue with `from __future__ import annotations`. What python version gives that issue so we can document it?

Reply:
> Also, note that the code does handle the possibility of other being something completely different than `ContextValue`

Yeah, I was thinking about that one. But is the usage intent to allow the comparison with non-`ContextValue` types? It would create a bit of ambiguity e.g. is it compared as `ContextValue(other)`?. This way at least the user is encouraged to check the implementation and make a specific type checker independent of that.

Reply:
> That would be an issue with from __future__ import annotations. What python version gives that issue so we can document it?

Well, IIUIC, every single one of them. When this bit of code is spotted by the Python interpreter, there is no object called `ContextValue` yet, because it's being defined. This is not about Python versions, it's natural, `ContextValue` is not yet complete and therefore it cannot be referenced by code like this. it's exactly the same situation as the following:

```python
class Foo:
    A = 1
    B = Foo.A
```

Python should throw this back at you with "name 'Foo' is not defined" on line 3 (`B = Foo.A`).

>> Also, note that the code does handle the possibility of other being something completely different than ContextValue

> Yeah, I was thinking about that one. But is the usage intent to allow the comparison with non-ContextValue types? It would create a bit of ambiguity e.g. is it compared as ContextValue(other)?. This way at least the user is encouraged to check the implementation and make a specific type checker independent of that.

I don't know what the intent was, I just see the code, and the code clearly accepts situations where `other` can be something else than `ContextValue`. It even makes a special branch for comparing other types with `ContextValue`. Both the annotation and code should be aligned, therefore, as I see it, we should either change the annotation to allow wider inputs or make the non-`ContextValue` case a hard error. If the annotation states "only `ContextValue` allowed", returning `False` for other values makes the error silently disappear.

Reply:
On my side, `python 3.11` + `PyCharm`, there is no error as long as `__future__` is included. I am not sure if that was introduced in `3.10` or `3.9`. Will have to check around on that.

> Both the annotation and code should be aligned, therefore, as I see it, we should either change the annotation to allow wider inputs or make the non-ContextValue case a hard error. If the annotation states "only ContextValue allowed", returning False for other values makes the error silently disappear.

Indeed we should get a clarification on this. It is particularly a problem if you are comparing `neq` and such.

Reply:
> On my side, python 3.11 + PyCharm, there is no error as long as __future__ is included. I am not sure if that was introduced in 3.10 or 3.9. Will have to check around on that.

Ah, got it, I misunderstood. I believe it should be possible to enable the postponed evaluation of annotations since Python 3.7.